query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -351,32 +398,57 @@ impl ConjoiningClauses {
-                            for (val, acc) in r.into_iter().zip(accumulated_types_for_columns.iter_mut()) {
+                            // TODO: don't accumulate twice.
+                            let mut vals = Vec::with_capacity(expected_width);
+                            let mut skip: Option<EmptyBecause> = None;
+                            for (col, pair) in cols.into_iter().zip(template.iter()) {
+                                // Now we have (val, Option<(name, known_types)>). Silly,
+                                // but this is how we iter!
+                                // Convert each item in the row.
+                                // If any value in the row is impossible, then skip the row.
+                                // If all rows are impossible, fail the entire CC.
+                                if let &Some(ref pair) = pair {
+                                    match self.typed_value_from_arg(schema, &pair.0, col, pair.1) {
+                                        ValueConversionResult::Ok(tv) => vals.push(tv),
+                                        ValueConversionResult::Err(e) => bail!(e),
+                                        ValueConversionResult::Impossible(because) => {
+                                            // Skip this row. It cannot produce bindings.
+                                            skip = Some(because);
+                                            break;
+                                        },
+                                    }
+                                }
+                            }
+
+                            if skip.is_some() {
+                                // Skip this row and record why, in case we skip all.
+                                skipped_all = skip;
+                                continue;
+                            }
+
+                            // Accumulate the values into the matrix and the types into the type set.
+                            for (val, acc) in vals.into_iter().zip(accumulated_types_for_columns.iter_mut()) {
                                 let inserted = acc.insert(val.value_type());
                                 if inserted && !acc.is_unit() {
                                     // Heterogeneous types.
                                     bail!(ErrorKind::InvalidGroundConstant);
                                 }
                                 matrix.push(val);
                             }
+
                         },
                         _ => bail!(ErrorKind::InvalidGroundConstant),
                     }
                 }
 
+                // Do we have rows? If not, the CC cannot succeed.
+                if matrix.is_empty() {
+                    // We will either have bailed or will have accumulated *something* into the matrix,
+                    // so we can safely unwrap here.
+                    self.mark_known_empty(skipped_all.expect("we skipped for a reason"));
```

Comment:
nit: "we skipped all rows for a reason".  The error message probably won't include a stack, so let's make the messages distinct.